In [1]:
import numpy as np
import pandas as pd

In [7]:
file=pd.read_excel(open('adult_salary_dataset.xlsx','rb'))
file.isnull().sum()

Age                0
Workclass          0
fnlwgt             0
education          0
eduction_no        1
status             0
occupation         0
relationship       0
race               0
sex                0
capital_gain       0
capital_loss       0
hoursperweek       0
native country     0
salary excedded    0
dtype: int64

There is one null value in education_no we will remove the whole row as it will not reduce the models generalization.

In [8]:
file=file.dropna()

In [9]:
file.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26444 entries, 1 to 26444
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Age              26444 non-null  int64  
 1   Workclass        26444 non-null  object 
 2   fnlwgt           26444 non-null  int64  
 3   education        26444 non-null  object 
 4   eduction_no      26444 non-null  float64
 5   status           26444 non-null  object 
 6   occupation       26444 non-null  object 
 7   relationship     26444 non-null  object 
 8   race             26444 non-null  object 
 9   sex              26444 non-null  object 
 10  capital_gain     26444 non-null  int64  
 11  capital_loss     26444 non-null  int64  
 12  hoursperweek     26444 non-null  int64  
 13  native country   26444 non-null  object 
 14  salary excedded  26444 non-null  object 
dtypes: float64(1), int64(5), object(9)
memory usage: 3.2+ MB


In [5]:
file.isnull().sum()

Age                0
Workclass          0
fnlwgt             0
education          0
eduction_no        0
status             0
occupation         0
relationship       0
race               0
sex                0
capital_gain       0
capital_loss       0
hoursperweek       0
native country     0
salary excedded    0
dtype: int64

There are no null values as we have removed the null values through dropna method.

In [10]:
file.describe()

Age        fnlwgt   eduction_no  capital_gain  capital_loss  \
count  26444.000000  2.644400e+04  26444.000000  26444.000000  26444.000000   
mean      38.612124  1.897716e+05     10.085161   1094.879406     86.486122   
std       13.676085  1.052112e+05      2.564187   7507.965257    401.207089   
min       17.000000  1.228500e+04      1.000000      0.000000      0.000000   
25%       28.000000  1.179630e+05      9.000000      0.000000      0.000000   
50%       37.000000  1.783840e+05     10.000000      0.000000      0.000000   
75%       48.000000  2.369522e+05     12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06     16.000000  99999.000000   4356.000000   

       hoursperweek  
count  26444.000000  
mean      40.411700  
std       12.299489  
min        1.000000  
25%       40.000000  
50%       40.000000  
75%       45.000000  
max       99.000000

Here we can see we have the  quartiles value, mean, standard_deviation and min-max values of all the features having int_64 datatype.

In [13]:
x=file.iloc[:,[1,3,5,6,7,8,9]].values
y=file.iloc[:,13].values

In [14]:
from sklearn.preprocessing import OneHotEncoder as ohe, LabelEncoder as le
y=le().fit_transform(y)
y

array([39, 39, 39, ..., 39, 39, 39])

In [16]:
np.unique(y)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41])

So we have 42 different classes in whole array 'y'. We have labelencoded our output.

In [19]:
y=y.reshape(-1,1)
y.shape

(26444, 1)

Now we are reshaping our array in (26444,1) means single column and 26444 rows.

In [44]:
y=ohe().fit_transform(y)
y

<26444x672 sparse matrix of type '<class 'numpy.float64'>'
	with 8885184 stored elements in Compressed Sparse Row format>

This is lazy loading i.e. the memory is assigned to the array but we can't see the output. It is used for the faster implementation of the code.

In [45]:
y.todense()

matrix([[0., 1., 1., ..., 1., 1., 0.],
        [0., 1., 1., ..., 1., 1., 0.],
        [0., 1., 1., ..., 1., 1., 0.],
        ...,
        [0., 1., 1., ..., 1., 1., 0.],
        [0., 1., 1., ..., 1., 1., 0.],
        [0., 1., 1., ..., 1., 1., 0.]])

Here we can see we are getting a matrix with todense method but we need array so we will use toarray() method.

In [46]:
y=y.toarray()

In [49]:
x=pd.DataFrame(x)
x=x.apply(le().fit_transform)
x

0   1  2   3  4  5  6
0      6   9  2   4  0  4  1
1      4  11  0   6  1  4  1
2      4   1  2   6  0  2  1
3      4   9  2  10  5  2  0
4      4  12  2   4  5  4  0
...   ..  .. ..  .. .. .. ..
26439  4  14  2  10  0  4  1
26440  4  11  2   7  0  4  1
26441  5  14  2  10  0  4  1
26442  4  15  4  12  3  4  0
26443  4  11  4   1  1  4  0

[26444 rows x 7 columns]

Now we will take all the input columns and apply labelencoder to it by .apply() method.

In [52]:
x=ohe().fit_transform(x).toarray()
x.shape

(26444, 240)

Here we will onehot encode the whole dataset and we will notice that number of columns are increased.

In [53]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest=train_test_split(x,y,test_size=0.2)

Now we will split our dataset into training and testing set of 80/20.

In [54]:
from sklearn.neighbors import KNeighborsClassifier as knn
KNN=knn(n_neighbors=3)
KNN.fit(xtrain,ytrain)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                     weights='uniform')

We are using Knn algo here and fitting the dataset in it. By default it is using Euclidean distance. 

In [55]:
print(KNN.score(xtest,ytest))

0.8844772168651919


The score we are getting here shows the accuracy of our model.